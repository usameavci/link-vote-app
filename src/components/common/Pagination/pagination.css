.common-pagination {
  grid-gap: 0.5rem;
  display: inline-grid;
  grid-auto-flow: column;
}

.common-pagination-page {
  padding: 0.5rem;
  cursor: pointer;
  min-width: 38px;
  text-align: center;
  border: 2px solid transparent;
}

.common-pagination-page:hover {
  border-color: #ddd;
}

.common-pagination-page.common-pagination-page--active {
  border-color: #000;
}
