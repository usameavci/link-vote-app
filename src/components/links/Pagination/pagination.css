.link-pagination {
  display: flex;
  margin-top: 2rem;
  align-items: center;
  justify-content: center;
}
