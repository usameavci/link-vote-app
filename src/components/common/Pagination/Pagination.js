import get from "lodash/get";
import map from "lodash/map";
import ceil from "lodash/ceil";
import reduce from "lodash/reduce";
import classNames from "classnames";
import uniqueId from "lodash/uniqueId";
import toSafeInteger from "lodash/toSafeInteger";
import React, { useEffect, useState } from "react";

import "./pagination.css";

const _createRange = (props, total, perPage = 5, current = 1, itemLimit = 3) => {
  const pageCount = ceil(total / perPage);

  current = current <= 0 ? 1 : current;
  current = current > pageCount ? pageCount : current;

  const range = [];
  const lowerLimit = current - itemLimit > 1 ? current - itemLimit : 1;
  const upperLimit = current + itemLimit < pageCount ? current + itemLimit : pageCount;

  if (1 < current) {
    range.push({
      key: uniqueId(),
      number: current - 1,
      label: props.prevLabel || "<",
      active: false,
    });
  }

  for (let i = lowerLimit; i <= upperLimit; i++) {
    range.push({
      key: uniqueId(),
      number: i,
      label: i,
      active: i === current,
    });
  }

  if (current < pageCount) {
    range.push({
      key: uniqueId(),
      number: current + 1,
      label: props.nextLabel || ">",
      active: false,
    });
  }

  return range;
};
const _getRangeParams = (props) =>
  reduce(
    ["total", "perPage", "current", "itemLimit"],
    (acc, curr) => acc.push(toSafeInteger(get(props, curr))) && acc,
    []
  );

const Pagination = (props) => {
  const rangeParams = _getRangeParams(props);
  const [range, setRange] = useState(_createRange(props, ...rangeParams));

  const handleOnClick = (page) => {
    const rangeParams = _getRangeParams({ ...props, current: page.number });
    setRange(_createRange(props, ...rangeParams));

    if (props.onChange && typeof props.onChange === "function") {
      props.onChange(page);
    }
  };

  const handleRangeChange = () => {
    setRange(_createRange(props, ...rangeParams));
  };

  useEffect(handleRangeChange, [props]);

  return (
    <div className="common-pagination">
      {map(range, (page) => (
        <div
          className={classNames("common-pagination-page", {
            "common-pagination-page--active": page.active,
          })}
          key={page.key}
          onClick={() => handleOnClick(page)}
        >
          {page.label}
        </div>
      ))}
    </div>
  );
};

export default Pagination;
