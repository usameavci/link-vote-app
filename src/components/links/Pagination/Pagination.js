import React from "react";
import { withRouter } from "react-router-dom";

import Url from "../../../utils/Url";
import CommonPagination from "../../common/Pagination/Pagination";

import "./pagination.css";

const Pagination = ({ pagination, history, location }) => {
  const url = new Url(location);

  const handleOnChange = (page) => {
    if (page.number && page.number !== 1) {
      url.setParam("page", page.number);
    } else {
      url.deleteParam("page");
    }
    history.push({ pathname: location.pathname, search: url.toString({ full: true }) });
  };

  return (
    <div data-testid="link-pagination" className="link-pagination">
      {pagination && pagination.count > 1 ? (
        <CommonPagination
          total={pagination.total}
          current={url.getParam("page")}
          perPage={pagination.perPage}
          itemLimit={3}
          onChange={handleOnChange}
          prevLabel={<i className="fa fa-chevron-left" />}
          nextLabel={<i className="fa fa-chevron-right" />}
        />
      ) : (
        ""
      )}
    </div>
  );
};

export default withRouter(Pagination);
