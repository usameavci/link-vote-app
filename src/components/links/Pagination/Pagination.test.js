import React from "react";
import renderer from "react-test-renderer";
import { render } from "@testing-library/react";
import { BrowserRouter as Router } from "react-router-dom";

import Pagination from "./Pagination";

describe("Index page component tests", () => {
  it("renders without crashing", () => {
    const component = (
      <Router>
        <Pagination />
      </Router>
    );

    const { getByTestId } = render(component);

    expect(getByTestId("link-pagination")).toBeInTheDocument();
  });

  it("matches snapshots", () => {
    const component = (
      <Router>
        <Pagination />
      </Router>
    );

    const tree = renderer.create(component);

    expect(tree.toJSON()).toMatchSnapshot();
  });
});
