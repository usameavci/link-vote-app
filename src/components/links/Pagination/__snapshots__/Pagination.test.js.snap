// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Index page component tests matches snapshots 1`] = `
<div
  className="link-pagination"
  data-testid="link-pagination"
>
  
</div>
`;
